Added try/catch for TypeDictionary errors in Lint code

PunkPun opened the pull request:
supersedes #20080

PunkPun commented:
See the lint

Mailaender commented on OpenRA.Mods.Common/Lint/CheckDefaultVisibility.cs:
```diff
@@ -33,24 +33,31 @@ void Run(Action<string> emitError, Ruleset rules)
 				}
+				catch (InvalidOperationException e)
+				{
+					emitError($"{e.Message} (Actor type `{actorInfo.Key}`)\n{e.StackTrace}");
```
Good idea, but I suggest avoiding the complete stacktrace. It is hard to read, especially when multiple errors occur.

Mailaender replied:
Also the errors above start with `Actor type {actorInfo.Key}` instead of ending with it.

PunkPun replied:
Sure, I had the same idea of removing stack traces after making this PR. It's not useful for modders.

And as for where Actor type is, I've seen it in errors be either in the middle or end of the message. But I don't mind putting it at start

PunkPun replied:
removed stack trace

abcdefg30 commented on OpenRA.Mods.Common/Lint/LintBuildablePrerequisites.cs:
```diff
@@ -33,14 +33,28 @@ void Run(Action<string> emitError, Ruleset rules)
+				{
+					var bi = actorInfo.Value.TraitInfoOrDefault<BuildableInfo>();
+					if (bi != null)
+					{
+						foreach (var prereq in bi.Prerequisites)
+						{
+							if (!prereq.StartsWith("~disabled"))
+							{
+								if (!providedPrereqs.Contains(prereq.Replace("!", "").Replace("~", "")))
+									emitError($"Buildable actor {actorInfo.Key} has prereq {prereq} not provided by anything.");
+							}
+						}
+					}
```
Style nit: Remove the braces around single-line statements.

PunkPun replied:
where

PunkPun replied:
Where? `Catch` requires braces

abcdefg30 replied:
You added extra braces around `if (bi != null)` and the foreach.

PunkPun replied:
yes, because it is not a single line statement, thus I fixed it for not following our convention

abcdefg30 replied:
We only have `emitError($"Buildable actor {actorInfo.Key} has prereq {prereq} not provided by anything.");` inside. We do it like this everywhere else in the code base.

PunkPun replied:
I've been made aware that we do not. I was specifically asked to fix the same style issue multiple tiumes. Also I haven't seen this anywhere in the code base but here

PunkPun committed:
Added try/catch for TypeDictionary errors in Lint code

TypeDictionary errors are very hard for modders to debug as they don't mention which actor is causing the error

PunkPun commented:
removed test case

abcdefg30 commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/bba54d20ea7be6afb5727a0c347b0defc5c2fd0f...57d5cfbe54c12448af35146070e577cea65b5803)
